### ISSUE TITLE
Non-zero exit code support

### ISSUE THREAD

**Issue by kentcb** [JordanMarr/FSharp.SystemCommandLine#4]
Hello,

Thanks for this library. I've been playing with it and it's a joy to use from F# compared to using System.CommandLine directly, which is entirely the point I know!

Do you have plans to support non-zero exit codes? A handler that returns `int` or `Task<int>` should allow for the exit code to be controlled, but it doesn't currently work because there are no overloads of `[Root]CommandBuilder.Run` that have `int` or `Task<int>` as their `'Output`.

Similar question for `InvocationContext`, though I expect the above is a simpler jump-off point.

**Comment by JordanMarr** [comment]
Thanks for the feedback! 

I just published v0.5 which will give you the ability to return a `Task<int>` from your `rootCommand` and `command` handler functions.

Admittedly I am not familiar with the `InvocationContext`.
Is the idea that you create a command by implementing `ICommandHandler` so that you can use DI and constructor injection?

**Comment by kentcb** [comment]
Wow, thanks for the quick turnaround on exit code support - will check it out later today.

Yeah, I'm no expert in `System.CommandLine` - only just started using it the other day. It seems like their [official docs](https://docs.microsoft.com/en-us/dotnet/standard/commandline/) are lagging behind the current feature set a bit. I found out about this via [this announcement](https://github.com/dotnet/command-line-api/issues/1537) that discusses `InvocationContext` and other "injectables". Tbh, I'm not sure important it would be to support these (though `CancellationToken` seems like a particularly useful one), nor how the explosion of permutations would impact the implementation.

**Comment by baronfel** [comment]
Good insight about the number of permutations: that's been a key design constraint in the implementation of Set handler. The thought was that past 16 parameters folks were reaching towards source generator or other codegen territory. The actual implementation of an ICommandHandler is really mechanical if all you're doing is binding to options or arguments.

The 'default' bindable types like ParseResult, InvocationContext, CancellationToken et al are spelled out, but our position is that most folks shouldn't have to care about InvocationContext specifically - it mostly is oriented towards Middleware authors. Being able to return an int from your handler is supposed to cover the main thing that folks do with the InvocationContext - set the exit code.

**Comment by kentcb** [comment]
@JordanMarr Just gave this a quick try and noticed I have to return `Task<int>` rather than just `int`, even if I have no need for asynchrony. This is as you described above, but guess I'm wondering whether you should also allow the latter - S.CL certainly does, but can appreciate it means translating `unit` into `Action` and anything else into `Func`

**Comment by JordanMarr** [comment]
Of the 34 `Command` `SetHandler` overloads that take `Func`, they all have a return type of `Task`.
So I don't see a way to implement this. How do you normally do this?

**Comment by JordanMarr** [comment]
@kentcb 
Good news: you can now inject stuff into your handler function in v0.6!

[v0.6 Release Info](https://github.com/JordanMarr/FSharp.SystemCommandLine/releases/tag/v0.6.0-beta)

**Comment by JordanMarr** [comment]
@baronfel , @kentcb:
[v0.7](https://github.com/JordanMarr/FSharp.SystemCommandLine/releases/tag/v0.7.0-beta) allows you to simply return an `int` status code from your handler:

```F#
module Program

open FSharp.SystemCommandLine

let app (words: string array, separator: string option) =
    let separator = separator |> Option.defaultValue ", "
    System.String.Join(separator, words) |> printfn "Result: %s"
    0
    
[<EntryPoint>]
let main argv = 
    let words = Input.Option(["--word"; "-w"], Array.empty, "A list of words to be appended")
    let separator = Input.OptionMaybe(["--separator"; "-s"], "A character that will separate the joined words.")

    rootCommand argv {
        description "Appends words together"
        inputs (words, separator)
        setHandler app
    }
```

**Comment by kentcb** [comment]
Hmm, unfortunately this works until you have two subcommands that both return plain old ints, which results in this (without even providing any arguments to my app):

```
Unhandled exception. System.ArgumentException: An item with the same key has already been added. Key: Visualizations
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.CommandLine.Parsing.StringExtensions.ValidTokens(Command command)
   at System.CommandLine.Parsing.StringExtensions.Tokenize(IReadOnlyList`1 args, CommandLineConfiguration configuration)
   at System.CommandLine.Parsing.Parser.Parse(IReadOnlyList`1 arguments, String rawInput)
   at FSharp.SystemCommandLine.CommandBuilders.RootCommandBuilder`17.Run[Inputs](CommandSpec`2 spec)
   at Program.main(String[] argv) in C:\Users\Kent\Repository\medical_history\Visualizations\Program.fs:line 92
```

**Comment by JordanMarr** [comment]
Good catch. Bug is fixed in [v0.7.1](https://www.nuget.org/packages/FSharp.SystemCommandLine/0.7.1-beta).

**Comment by kentcb** [comment]
Fantastic! Thanks @JordanMarr. Closing this out.
